File: 1140-Stone Game II/solution.java
class Solution {
    public int stoneGameII(int[] piles) {
        int n = piles.length;
        int[] suffixSums = new int[n + 1];
        suffixSums[n - 1] = piles[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            suffixSums[i] = suffixSums[i + 1] + piles[i];
        }

        int[][] dp = new int[n][n + 1];

        for (int i = n - 1; i >= 0; i--) {
            for (int m = 1; m <= n; m++) {
                if (i + 2 * m >= n) {
                    dp[i][m] = suffixSums[i];
                } else {
                    for (int x = 1; x <= 2 * m; x++) {
                        int opponentScore = dp[i + x][Math.max(x, m)];
                        int score = suffixSums[i] - opponentScore;
                        dp[i][m] = Math.max(dp[i][m], score);
                    }
                }
            }
        }

        return dp[0][1];
    }
}

File: 1140-Stone Game II/solution.py
class Solution:
    def stoneGameII(self, piles):
        n = len(piles)
        suffix_sums = [0] * (n + 1)
        suffix_sums[n - 1] = piles[n - 1]
        for i in range(n - 2, -1, -1):
            suffix_sums[i] = suffix_sums[i + 1] + piles[i]

        dp = [[0] * (n + 1) for _ in range(n)]

        for i in range(n - 1, -1, -1):
            for m in range(1, n + 1):
                if i + 2 * m >= n:
                    dp[i][m] = suffix_sums[i]
                else:
                    for x in range(1, 2 * m + 1):
                        opponent_score = dp[i + x][max(x, m)]
                        score = suffix_sums[i] - opponent_score
                        dp[i][m] = max(dp[i][m], score)
        
        return dp[0][1]